tests/tools/test_validate_tool_call.py

```diff
@@ -0,0 +1,184 @@
+        )
+        self.assertTrue(result)
+        self.assertEqual(message, "All required tools called correctly")
+
+    def test_only_forbidden_tools_called(self):
+        result, message = validate_tool(
+            required_tools=[],
+            actual_tool_calls=["tool3"],
+            optional_tools=[],
+            forbidden_tools=["tool3"],
+        )
+        self.assertFalse(result)
+        self.assertEqual(message, "Forbidden tool called: tool3")
+
+    def test_mixed_tools_called(self):
+        result, message = validate_tool(
+            required_tools=["tool1"],
+            actual_tool_calls=["tool1", "tool2", "tool3"],
+            optional_tools=["tool2"],
+            forbidden_tools=["tool3"],
+        )
+        self.assertFalse(result)
+        self.assertEqual(message, "Forbidden tool called: tool3")
+
+    def test_repeated_required_tools(self):
+        result, message = validate_tool(
+            required_tools=["tool1", "tool1"],
+            actual_tool_calls=["tool1", "tool1"],
+            optional_tools=[],
+            forbidden_tools=[],
+        )
+        self.assertTrue(result)
+        self.assertEqual(message, "All required tools called correctly")
+
+    def test_repeated_forbidden_tools(self):
+        result, message = validate_tool(
+            required_tools=["tool1"],
+            actual_tool_calls=["tool1", "tool3", "tool3"],
+            optional_tools=[],
+            forbidden_tools=["tool3"],
+        )
+        self.assertFalse(result)
+        self.assertEqual(message, "Forbidden tool called: tool3")
+
+    def test_overrepeated_tools(self):
+        result, message = validate_tool(
+            required_tools=["tool1", "tool2", "tool3"],
+            actual_tool_calls=["tool1", "tool2", "tool2", "tool2", "tool3"],
+            optional_tools=[],
+            forbidden_tools=[],
+        )
+        self.assertTrue(result)
+        self.assertEqual(message, "All required tools called correctly")
+
+    def test_overrepeated_tools2(self):
+        result, message = validate_tool(
+            required_tools=["tool1", "tool2", "tool3"],
+            actual_tool_calls=["tool1", "tool2", "tool3", "tool3", "tool3"],
+            optional_tools=[],
+            forbidden_tools=[],
+        )
+        self.assertTrue(result)
+        self.assertEqual(message, "All required tools called correctly")
+
+    def test_overrepeated_tools3(self):
+        result, message = validate_tool(
+            required_tools=["tool1", "tool2", "tool3"],
+            actual_tool_calls=[
+                "tool1",
+                "tool1",
+                "tool1",
+                "tool2",
+                "tool2",
+                "tool3",
+                "tool3",
+            ],
+            optional_tools=[],
+            forbidden_tools=[],
+        )
+        self.assertTrue(result)
+        self.assertEqual(message, "All required tools called correctly")
+
+
+if __name__ == "__main__":
+    unittest.main()
```

Comment:
Could you add a small test for the function that calls the API (mocking it) ?

Reply:
adding it for async case. maybe we dont need sync case then